webapp-django/crashstats/crashstats/models.py

```diff
@@ -161,7 +161,96 @@ def inner(*args, **kwargs):
     return inner
 
 
+def memoize(function):
+    """Decorator for model methods to cache in memory or the filesystem
+    using CACHE_MIDDLEWARE and/or CACHE_MIDDLEWARE_FILES Django config"""
+
+    @functools.wraps(function)
+    def memoizer(*args, **kwargs):
+
+        def get_cached_result(key):
+            result = cache.get(key)
+            if result is not None:
+                logger.debug("CACHE HIT %s" % stringified_args)
+                return result
+
+            # Didn't find key in middelware_cache, so try filecache
+            if settings.CACHE_MIDDLEWARE_FILES:
+                # but is it fresh enough?
+                age = time.time() - os.stat(cache_file)[stat.ST_MTIME]
+                if age > instance.cache_seconds:
+                    logger.debug("CACHE FILE TOO OLD")
+                    os.remove(cache_file)
+                else:
+                    logger.debug("CACHE FILE HIT %s" % stringified_args)
+                    if instance.expect_json:
+                        return json.load(open(cache_file))
+                    else:
+                        return open(cache_file).read()
+
+            # Didn't find our values in the cache
+            return None
+
+        def get_cache_filename(key):
+            root = settings.CACHE_MIDDLEWARE_FILES
+            if isinstance(root, bool):
+                cache_file = os.path.join(
+                    settings.ROOT,
+                    'models-cache'
+                )
+            else:
+                cache_file = root
+
+            cache_file = os.path.join(
+                cache_file,
+                classname,
+                key
+            )
+            if instance.expect_json:
+                cache_file = os.path.join(
+                    cache_file,
+                    '%s.json' % key
+                )
+            else:
+                cache_file = os.path.join(
+                    cache_file,
+                    '%s.dump' % key
+                )
```

Comment:
It was repeated twice. I thought I was copying existing functionality, but probably missed something.

I like your suggested version much better.